fuse client, fscrypt, test: Implement and create tests for S_ENCRYPTED in inode i_flags

salieri11 opened the pull request:
This PR adds test for `S_ENCRYPTED` bit in the `i_flags` field of `Inode`.
The test implements 2 quering methods: using `FS_IOC_GETFLAGS` and `STATX_ATTR_ENCRYPTED`

Fixes: https://tracker.ceph.com/issues/64129

<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

chrisphoffman commented on src/client/Inode.h:
```diff
@@ -178,6 +178,9 @@ struct Inode : RefCountedObject {
   uint64_t effective_size() const;
   void set_effective_size(uint64_t size);
 
+  // this methods returns true iff inode is de facto ecrypted.
```
nit: spelling

chrisphoffman commented on src/client/fuse_ll.cc:
```diff
@@ -1122,6 +1122,20 @@ static void fuse_ll_ioctl(fuse_req_t req, fuse_ino_t ino,
+      }
+
+      Fh *fh = (Fh*)fi->fh;
+      cfuse->client->ll_handle_ioctl(fh->inode.get(), FS_IOC_GETFLAGS, &file_attr);
```
We're already handling ioctl cmds here no need to do it again later in `ll_handle_ioctl`.

Can you make it call something like `get_inode_flags`?

salieri11 replied:
Right, now I see that client just needs to fill the flags. Fixed.
I wonder whether the client needs just to return i_flags or to set explicitly bits that we support right now (e.g. FS_ENCRYPT_FL) ?

chrisphoffman commented on src/libcephfs.cc:
```diff
@@ -2420,3 +2420,11 @@ extern "C" void ceph_free_snap_info_buffer(struct snap_info *snap_info) {
   free(snap_info->snap_metadata);
 }
+
+extern "C" int ceph_ioctl(struct ceph_mount_info *cmount, int fd, int command, int* file_attr_out)
```
I'm not sure we want to expose ioctls at the API level. I've been creating helper functions that call the defined client functions instead.

See `ceph_add_fscrypt_key`, `ceph_remove_fscrypt_key`, `ceph_get_fscrypt_key_status` etc.

If using the example from other comment, having it call `get_inode_flags` would follow that pattern.

salieri11 replied:
I wonder if the requirment is to query i_flag status vis IOCTL, we should implement the interface? Otherwise, why should we support FS_IOC_GETFLAGS is there is no IOCTL support in the interface?

chrisphoffman reviewed (commented):
I'm having trouble getting this to compile.

```
error: narrowing conversion of ‘2148034049’ from ‘long unsigned int’ to ‘int’ [-Wnarrowing]
  138 | #define FS_IOC_GETFLAGS _IOR('f', 1, long)
```

What environment are you testing this on that compiles?

Here's some further info that seems to be related: https://lore.kernel.org/all/20131126200559.GH20559@hall.aurel32.net/T/

salieri11 commented:
> I'm having trouble getting this to compile.
> 
> ```
> error: narrowing conversion of ‘2148034049’ from ‘long unsigned int’ to ‘int’ [-Wnarrowing]
>   138 | #define FS_IOC_GETFLAGS _IOR('f', 1, long)
> ```
> 
> What environment are you testing this on that compiles?
> 
> Here's some further info that seems to be related: https://lore.kernel.org/all/20131126200559.GH20559@hall.aurel32.net/T/

I am building both on Ubuntu 24 (arm version), and on vossi04 (i guess its debian), with no errors, using do_cmake.sh
What is your env and how do you build the code?

chrisphoffman commented:
> I am building both on Ubuntu 24 (arm version), and on vossi04 (i guess its debian), with no errors, using do_cmake.sh What is your env and how do you build the code?

Fedora and do_cmake as well. I'll try to take a look more at this this week.

salieri11 commented:
> > I am building both on Ubuntu 24 (arm version), and on vossi04 (i guess its debian), with no errors, using do_cmake.sh What is your env and how do you build the code?
> 
> Fedora and do_cmake as well. I'll try to take a look more at this this week.
I changed it to:
#define FS_IOC_GETFLAGS _IOR('f', 1, unsigned long)
This is legit fix to use unsigned

> > I am building both on Ubuntu 24 (arm version), and on vossi04 (i guess its debian), with no errors, using do_cmake.sh What is your env and how do you build the code?
> 
> Fedora and do_cmake as well. I'll try to take a look more at this this week.

I changed it to 
`#define FS_IOC_GETFLAGS _IOR('f', 1, int32_t)`
usually its good enough for these cases, just to use fixed width integer types

salieri11 commented on src/client/Client.cc:
```diff
@@ -18352,6 +18358,27 @@ mds_rank_t Client::_get_random_up_mds() const
+  if (!file_attr_out)
+      return -CEPHFS_EINVAL;
+
+   *file_attr_out = 0; 
```
done

salieri11 commented on src/client/Client.cc:
```diff
@@ -8966,6 +8966,12 @@ void Client::fill_statx(Inode *in, unsigned int mask, struct ceph_statx *stx)
+  if (in->is_encrypted()) {
+    stx->stx_attributes |= STATX_ATTR_ENCRYPTED;
+  } else {
+    stx->stx_attributes &= ~STATX_ATTR_ENCRYPTED;
```
Well not for now, but in general, I prefer to have clear semantics in such case, without assuming that the flag is clear by default (and not having some garbage that may set the flag to ENCRYPTED)

salieri11 commented on src/client/Inode.h:
```diff
@@ -193,6 +196,16 @@ struct Inode : RefCountedObject {
+    if (en) {
+      i_flags |= S_ENCRYPTED;
+    } else {
+      i_flags &= (~S_ENCRYPTED);
```
Same as above:) For better semantics and defensive code.

chrisphoffman replied:
I see your reasoning on this. I'm concerned that if there's "garbage" in it, we should let it show instead of hiding the issue.

I'm looking at linux kernel ioctl.c and during 
`fileattr_fill_flags|fileattr_fill_xflags`

it sets value and doesn't worry about clearing.

salieri11 replied:
It makes sense for me now.I think that best will be to assert(false) if `is_fscrypt_enabled` is false but the S_ENCRYPTED bit is set. It wil let us catch the problem, instead of cleaning it.

chrisphoffman replied:
I like that, let's proceed with that change.

salieri11 commented on src/test/libcephfs/fscrypt.cc:
```diff
@@ -949,6 +945,79 @@ TEST(FSCrypt, RemoveBusyCreate) {
+  path.append(dir_path);
+  path.append("/");
+  path.append(file_path);
+  int fd1 = ceph_open(cmount, path.c_str(), O_RDWR|O_CREAT|O_TRUNC, 0600);
```
agree, changed to fd1 and fd2, for better clarity

chrisphoffman commented on src/test/libcephfs/fscrypt.cc:
```diff
@@ -949,6 +945,79 @@ TEST(FSCrypt, RemoveBusyCreate) {
 
   ceph_unlink(cmount, src_path.c_str());
   ceph_rmdir(cmount, dir_path.c_str());
+  ceph_unmount(cmount);
```
If this is needed perhaps we should do that for all tests. I'm fine with this staying.

chrisphoffman commented:
Merged into branch manually due to rebase cleanup work.

See commit here
https://github.com/ceph/ceph/pull/61137/commits/6d15f1d31762446903f8956b8f092b7adfcc3276

chrisphoffman commented:
This is a test link, pasting this
`https://github.com/chrisphoffman/ceph/commit/d8f6c2cfe248d3af786215eb7e567645c5655b74`

without code annotation https://github.com/chrisphoffman/ceph/commit/d8f6c2cfe248d3af786215eb7e567645c5655b74
